[libc++] Implement P3016R6: Resolve inconsistencies in begin/end for `valarray` and braced initializer lists

frederick-vs-ja opened the pull request:
This paper is implemented as a Defect Report against C++11 and later revisions.
- Inclusion of `<typeinfo>` in `<functional>` should be considered as a patch to C++ TR1 [N1836](https://wg21.link/N1836), while the related parts were merged into C++11.
- Changes of `initializer_list` should be considered as a patch to C++11 paper [N2672](https://wg21.link/n2672).
- Changes of `valarray`, `std::begin`, and `std::end` and availability of range access functions in `<valarray>` should be considered as a patch to C++11 paper [N2930](https://wg21.link/n2930).
- Inclusion of `<typeinfo>` in `<typeindex>` should be considered as a patch to C++11 paper [N2932](https://wg21.link/n2932).
- Changes of `std::cbegin`, `std::cend`, `std::rbegin`, `std::rend`, `std::crbegin`, and `std::crend` and the inclusion of `<initializer_list>` in `<iterator>` should be considered as a patch to [LWG2128](https://wg21.link/lwg2128), while LWG2128 is consistently considered to be a pure feature addition in C++14 by implementations.
- Changes of `std::data`, `std::empty`, and `std::size` should be considered as a patch to C++17 paper [N4280](https://wg21.link/n4280).
- Inclusion of `<initializer_list>` and `<typeinfo>` in `<any>` should be considered as a patch to C++17 paper [P0220R1](https://wg21.link/p0220r1).
- Changes of `std::ssize` should be considered as a patch to C++20 paper [P1227R2](https://wg21.link/p1227r2).
- Inclusion of `<initializer_list>` in `<functional>` should be considered as a patch to C++23 paper [P0288R9](https://wg21.link/p0288r9).
- Availability of range access functions in `<stacktrace>` should be considered as a patch to C++23 paper [P0881R7](https://wg21.link/p0881r7).
- Availability of range access functions in `<optional>` should be considered as a patch to C++26 paper [P3168R2](https://wg21.link/p3168r2).

Notes for backported and missing parts:
- Changes of `<stacktrace>` are not implemented because P0881R7 is not implemented. But I think P3016R6 should be considered complete, with note added.
- `constexpr` for member functions of `initializer_list` is only added since C++14, in the spirit of C++14 paper [N3471](https://wg21.link/n3471).
- Changes of inclusions and `valarray` are unconditionally implemented, which follows the current convention of libc++.

Fixes #171271.

frederick-vs-ja committed:
[libc++] P3016R6: Resolve inconsistencies in begin/end for `valarray` and braced initializer lists

This paper is implemented as a Defect Report against C++11 and later
revisions.
- Inclusion of `<typeinfo>` in `<functional>` should be considered as a
patch to C++ TR1 [N1836](https://wg21.link/N1836), while the related
parts were merged into C++11.
- Changes of `initializer_list` should be considered as a patch to C++11
paper [N2672](https://wg21.link/n2672).
- Changes of `valarray`, `std::begin`, and `std::end` and availability
of range access functions in `<valarray>` should be considered as a
patch to C++11 paper [N2930](https://wg21.link/n2930).
- Inclusion of `<typeinfo>` in `<typeindex>` should be considered as a
patch to C++11 paper [N2932](https://wg21.link/n2932).
- Changes of `std::cbegin`, `std::cend`, `std::rbegin`, `std::rend`,
`std::crbegin`, and `std::crend` and the inclusion of
`<initializer_list>` in `<iterator>` should be considered as a patch to
[LWG2128](https://wg21.link/lwg2128), while LWG2128 is consistently
considered to be a pure feature addition in C++14 by implementations.
- Changes of `std::data`, `std::empty`, and `std::size` should be
considered as a patch to C++17 paper [N4280](https://wg21.link/n4280).
- Inclusion of `<initializer_list>` and `<typeinfo>` in `<any>` should
be considered as a patch to C++17 paper
[P0220R1](https://wg21.link/p0220r1).
- Changes of `std::ssize` should be considered as a patch to C++20 paper
[P1227R2](https://wg21.link/p1227r2).
- Inclusion of `<initializer_list>` in `<functional>` should be
considered as a patch to C++23 paper
[P0288R9](https://wg21.link/p0288r9).
- Availability of range access functions in `<stacktrace>` should be
considered as a patch to C++23 paper
[P0881R7](https://wg21.link/p0881r7).
- Availability of range access functions in `<optional>` should be
considered as a patch to C++26 paper
[P3168R2](https://wg21.link/p3168r2).

Notes for backported and missing parts:
- Changes of `<stacktrace>` are not implemented because P0881R7 is not
implemented. But I think P3016R6 should be considered complete with note
added.
- `constexpr` for member functions of `initializer_list` is only added
since C++14, in the spirit of C++14 paper
[N3471](https://wg21.link/n3471).
- Changes of inclusions and `valarray` are unconditionally implemented,
which follows the current convention of libc++.

frederick-vs-ja committed:
[libc++] P3016R6: Resolve inconsistencies in begin/end for `valarray` and braced initializer lists

This paper is implemented as a Defect Report against C++11 and later
revisions.
- Inclusion of `<typeinfo>` in `<functional>` should be considered as a
patch to C++ TR1 [N1836](https://wg21.link/N1836), while the related
parts were merged into C++11.
- Changes of `initializer_list` should be considered as a patch to C++11
paper [N2672](https://wg21.link/n2672).
- Changes of `valarray`, `std::begin`, and `std::end` and availability
of range access functions in `<valarray>` should be considered as a
patch to C++11 paper [N2930](https://wg21.link/n2930).
- Inclusion of `<typeinfo>` in `<typeindex>` should be considered as a
patch to C++11 paper [N2932](https://wg21.link/n2932).
- Changes of `std::cbegin`, `std::cend`, `std::rbegin`, `std::rend`,
`std::crbegin`, and `std::crend` and the inclusion of
`<initializer_list>` in `<iterator>` should be considered as a patch to
[LWG2128](https://wg21.link/lwg2128), while LWG2128 is consistently
considered to be a pure feature addition in C++14 by implementations.
- Changes of `std::data`, `std::empty`, and `std::size` should be
considered as a patch to C++17 paper [N4280](https://wg21.link/n4280).
- Inclusion of `<initializer_list>` and `<typeinfo>` in `<any>` should
be considered as a patch to C++17 paper
[P0220R1](https://wg21.link/p0220r1).
- Changes of `std::ssize` should be considered as a patch to C++20 paper
[P1227R2](https://wg21.link/p1227r2).
- Inclusion of `<initializer_list>` in `<functional>` should be
considered as a patch to C++23 paper
[P0288R9](https://wg21.link/p0288r9).
- Availability of range access functions in `<stacktrace>` should be
considered as a patch to C++23 paper
[P0881R7](https://wg21.link/p0881r7).
- Availability of range access functions in `<optional>` should be
considered as a patch to C++26 paper
[P3168R2](https://wg21.link/p3168r2).

Notes for backported and missing parts:
- Changes of `<stacktrace>` are not implemented because P0881R7 is not
implemented. But I think P3016R6 should be considered complete with note
added.
- `constexpr` for member functions of `initializer_list` is only added
since C++14, in the spirit of C++14 paper
[N3471](https://wg21.link/n3471).
- Changes of inclusions and `valarray` are unconditionally implemented,
which follows the current convention of libc++.

Drive-by changes:
- Conditionally strengthen the constructor of `reverse_iterator` from
the underlying iterator, because many (un)conditionally `noexcept`
functions call that constructor now.

philnik777 commented on libcxx/include/__iterator/reverse_iterator.h:
```diff
@@ -89,7 +89,9 @@ class reverse_iterator
 
-  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX17 explicit reverse_iterator(_Iter __x) : __t_(__x), current(__x) {}
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX17 explicit reverse_iterator(_Iter __x)
+      _NOEXCEPT_(is_nothrow_constructible<_Iter, _Iter&>::value)
```
Where does this come from? Same below.

frederick-vs-ja replied:
This is a drive-by change explicitly mentioned in the PR description. I'll drop it if you think it shouldn't be in this PR.

philnik777 commented on libcxx/include/any:
```diff
@@ -106,9 +109,11 @@ namespace std {
 #  include <__verbose_abort>
+#  include <version>
+
+// standard-mandated includes
```
Please add where the requirement comes from.

philnik777 commented on libcxx/include/functional:
```diff
@@ -596,14 +598,18 @@ POLICY:  For non-variadic implementations, the number of arguments is limited
 #    include <__vector/vector.h>
 #  endif
+
+// standard-mandated includes
```
Same here. The required includes should be before any `_LIBCPP_REMOVE_TRANSITIVE_INCLUDES` as well.

philnik777 commented on libcxx/include/valarray:
```diff
@@ -786,6 +788,9 @@ public:
   typedef _Tp __result_type;
 
+  typedef _Tp* iterator;
+  typedef const _Tp* const_iterator;
```
Let's switch to using declarations here. Also, do we want to use `__wrap_iter` here?

frederick-vs-ja replied:
`__wrap_iter` seems reasonable. But switching to use it looks like a larger design change to me (at least it requires an ABI option).

philnik777 reviewed (changes requested):
I haven't looked at the tests yet.

frederick-vs-ja commented on libcxx/include/version:
```diff
@@ -303,6 +305,7 @@ __cpp_lib_void_t                                        201411L <type_traits>
 # define __cpp_lib_complex_udls                         201309L
 # define __cpp_lib_exchange_function                    201304L
 # define __cpp_lib_generic_associative_lookup           201304L
+# define __cpp_lib_initializer_list                     202511L
```
The FTM values are already determined in [[version.syn]](https://eel.is/c++draft/version.syn).

frederick-vs-ja commented on libcxx/include/version:
```diff
@@ -303,6 +305,7 @@ __cpp_lib_void_t                                        201411L <type_traits>
 # define __cpp_lib_complex_udls                         201309L
 # define __cpp_lib_exchange_function                    201304L
 # define __cpp_lib_generic_associative_lookup           201304L
+# define __cpp_lib_initializer_list                     202511L
```
I mentioned these in https://github.com/llvm/llvm-project/pull/136528#discussion_r2648045076.

philnik777 replied:
Please put that somewhere more visible. There is no guarantee that that patch will actually land.

frederick-vs-ja replied:
OK. Mentioned again in https://github.com/llvm/llvm-project/issues/105131#issuecomment-3713200034.

frederick-vs-ja committed:
Address some of @philnik777's review comments

- Add section labels to standard-mandated includes
- Reorder includes in `<functional>`
- Remove inclusion guard for `<__iterator/reverse_iterator.h>`
- Use `using` for new member types in `valarray`

frederick-vs-ja committed:
Address some of @philnik777's review comments

- Add section labels to standard-mandated includes
- Reorder includes in `<functional>`
- Remove inclusion guard for `<__iterator/reverse_iterator.h>`
- Use `using` for new member types in `valarray`

frederick-vs-ja committed:
Fix typo in `valarray`

frederick-vs-ja committed:
Fix typo in `valarray`

frederick-vs-ja committed:
Revert drive-by exception specification strengthening

frederick-vs-ja committed:
Revert drive-by exception specification strengthening

frederick-vs-ja committed:
Make `reverse_access` export `reverse_iterator`

frederick-vs-ja committed:
Make `reverse_access` export `reverse_iterator`

frederick-vs-ja commented on libcxx/include/optional:
```diff
@@ -272,6 +272,7 @@ namespace std {
 #  include <__functional/unary_function.h>
 #  include <__fwd/functional.h>
 #  include <__iterator/bounded_iter.h>
+#  include <__iterator/reverse_iterator.h>
```
Thanks. Now changed.

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

# Conflicts:
#	libcxx/docs/ReleaseNotes/22.rst

frederick-vs-ja committed:
Retarget LLVM 23

frederick-vs-ja committed:
Retarget LLVM 23

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

# Conflicts:
#	libcxx/include/__iterator/access.h
#	libcxx/include/__iterator/reverse_access.h
#	libcxx/include/valarray

frederick-vs-ja committed:
clang-format

frederick-vs-ja committed:
clang-format

ldionne reviewed (approved):
This LGTM w/ rebase onto `main`, adjusting the release notes and C++26 status, and green CI.

I agree that we shouldn't introduce `__wrap_iter` in this change, it's an ABI break whether we were exposing `::iterator` and `::const_iterator` previously.

frederick-vs-ja committed:
Retarget LLVM 24

frederick-vs-ja committed:
clang-format, again

frederick-vs-ja committed:
Merge branch 'main' into p3016r6

ldionne committed:
Merge branch 'main' into p3016r6

frederick-vs-ja committed:
Fix transitive include lists for `<functional>`

philnik777 commented on libcxx/include/optional:
```diff
@@ -275,6 +275,13 @@ namespace std {
 // [optional.syn]
 #  include <compare>
 
+// [iterator.range]
+#  include <__iterator/access.h>
+#  include <__iterator/data.h>
+#  include <__iterator/empty.h>
+#  include <__iterator/reverse_access.h>
+#  include <__iterator/size.h>
```
When did these mandatory includes get introduced? Right now these increase the size of `<optional>` by ~8%.

frederick-vs-ja replied:
These includes address [LWG4131](https://cplusplus.github.io/LWG/issue4131) which is meaningful only since C++26 (due to [P3168R2](https://wg21.link/p3168r2)). Do you think it would be better to include these headers only since C++26?

philnik777 replied:
Since we're in an umbrella header, I think so, yeah.

frederick-vs-ja replied:
Opened #214688.
